src/backend/utils/misc/queryenvironment.c

```diff
@@ -1524,113 +1402,282 @@ void ENRRollbackSubtransaction(SubTransactionId subid, QueryEnvironment *queryEn
+			break;
+	}
+
+	/*
+	 * lcons here takes the length of the list, so it is less efficient than lappend. 
+	 * However, we want the uncommitted_tup order to behave more like a stack to 
+	 * process ROLLBACK in the right order of operations. 
+	 */
+	*list_ptr = lcons(uncommitted_tup, *list_ptr);
 }
 
-/* Simple check for whether to use temp OID buffer given an existing OID. */
-bool useTempOidBufferForOid(Oid relId)
+/*
+ * Clean up all uncommitted tuple data from this ENR.
+ * 
+ * This is called on COMMIT, to wipe unnecessary uncommitted tuple data away
+ * and on ROLLBACK, once the changes have been rolled back. 
+ */
+static void
+ENRDeleteUncommittedTupleData(SubTransactionId subid, EphemeralNamedRelation enr)
 {
-	return sql_dialect == SQL_DIALECT_TSQL 
-		&& GetNewTempOidWithIndex_hook 
-		&& temp_oid_buffer_size > 0
-		&& get_ENR_withoid(currentQueryEnv, relId, ENR_TSQL_TEMP);
+	for (int i = 0; i < ENR_CATTUP_END; i++)
+	{
+		ListCell *lc;
+
+		foreach(lc, enr->md.uncommitted_cattups[i])
+		{
+			ENRUncommittedTuple uncommitted_tup = (ENRUncommittedTuple) lfirst(lc);
+
+			if (subid != InvalidSubTransactionId && uncommitted_tup->subid < subid)
+				continue;
+
+			heap_freetuple(uncommitted_tup->tup);
+			pfree(uncommitted_tup);
+
+			enr->md.uncommitted_cattups[i] = foreach_delete_current(enr->md.uncommitted_cattups[i], lc);
+		}
+		if (subid == InvalidSubTransactionId)
+		{
+			list_free(enr->md.uncommitted_cattups[i]);
+			enr->md.uncommitted_cattups[i] = NIL;
+		}
+	}
 }
 
 /*
- * Drop all the temp tables registered as ENR in the given query environment.
+ * Helper function - finish the ROLLBACK processing for a single uncommitted_tup entry.
  */
-void
-ENRDropTempTables(QueryEnvironment *queryEnv)
+static void
+ENRRollbackUncommittedTuple(QueryEnvironment *queryEnv, ENRUncommittedTuple uncommitted_tup)
 {
-	ListCell   *lc = NULL;
-	ObjectAddress object;
-	ObjectAddresses *objects;
+	Relation rel;
+	bool skip_cache_inval = false;
 
-	if (!queryEnv)
+	if (!OidIsValid(uncommitted_tup->catalog_oid) || !HeapTupleIsValid(uncommitted_tup->tup))
+	{
+		elog(WARNING, "Invalid temp table entries were found during ROLLBACK");
 		return;
+	}
 
-	objects = new_object_addresses();
-
-	/*
-	 * Loop through the registered ENRs to drop temp tables.
-	 */
-	foreach(lc, queryEnv->namedRelList)
+	if (uncommitted_tup->catalog_oid == IndexRelationId && uncommitted_tup->optype == ENR_OP_ADD)
 	{
-		EphemeralNamedRelation enr = (EphemeralNamedRelation) lfirst(lc);
+		/*
+		* We skip invalidating IndexRelationId on drop (ie committed table, index created in transaction) 
+		* because it will be taken care of when the index itself is dropped - otherwise we risk
+		* throwing an error because the entry is already wiped away.
+		*/
+		
+		Form_pg_index idx_form = (Form_pg_index) GETSTRUCT(uncommitted_tup->tup);
+		EphemeralNamedRelation tmp_enr = get_ENR_withoid(queryEnv, idx_form->indrelid, ENR_TSQL_TEMP);
+		if (tmp_enr)
```

Comment:
nit - we dont use `tmp_enr` anywhere else, you can just call `get_ENR_withoid()` directly.